include/pybind11/stl.h

```diff
@@ -136,6 +147,15 @@ struct map_caster {
     PYBIND11_TYPE_CASTER(Type,
                          const_name("Dict[") + key_conv::name + const_name(", ") + value_conv::name
                              + const_name("]"));
+
+private:
+    template <
+        typename T = Type,
+        enable_if_t<std::is_same<decltype(std::declval<T>().reserve(0)), void>::value, int> = 0>
```

Comment:
Is this meant to be a stand-in for "value has a .reserve() method?" ?

Might be good to add a terse comment:
```
// This will work only for `reserve()` with `void` return (all std containers).
```
(Is "all std containers" true?)

Reply:
"All" sounds like this works on non-set or non-map containers too. This is a fairly standard idiom for checking to see if a method exists, I don't think it needs a comment. Not against one, though.

Reply:
@rwgk Not all std::containers have reserve (std::map, std::set, std::forward list) etc.. don't.

Reply:
I realize, what I had in mind is that all existing `reserve()` have a `void` return. If we're not sure, simply:

```
// This will work only for `reserve()` with `void` return.
```

So people don't have to parse the complicated `enable_if` to figure that out.